concurrency/DurableLocalStoreManagerImpl.java

```diff
@@ -0,0 +1,287 @@
+     * 
+     * @return
+     */
+    protected DurableLocalStore newStore(String bucketName, boolean create) throws StorageNodeException {
+        return new DurableLocalJsonStore(sim, bucketName, create);
+    }
+
+    public DurableLocalStoreManagerImpl(StorageInstanceManager sim) {
+        this.sim = sim;
+        this.bucketDataMap = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public boolean putOrUpdate(String bucketName, IndexKey key, Item value) throws StorageNodeException {
+        final DurableLocalStore store = getBucketData(bucketName);
+        store.assertIfNotOpen("putOrUpdate");
+        return store.putOrUpdate(key, value);
+    }
+
+    @Override
+    public Item get(String bucketName, IndexKey key) throws StorageNodeException {
+        final DurableLocalStore store = getBucketData(bucketName);
+        store.assertIfNotOpen("get");
+        return store.get(key);
+    }
+
+    @Override
+    public void create(String bucketName) throws StorageNodeException {
+        DurableLocalStore dataStore = bucketDataMap.get(bucketName);
+
+        if (dataStore == null) {
+            synchronized (this) {
+                dataStore = bucketDataMap.get(bucketName);
+                if (dataStore == null) {
+                    dataStore = newStore(bucketName, true);
+                    bucketDataMap.put(bucketName, dataStore);
+                    return;
+                }
+            }
+        }
+    }
+
+    @Override
+    public void load(String bucketName) throws StorageNodeException {
+        DurableLocalStore dataStore = bucketDataMap.get(bucketName);
+
+        if (dataStore == null) {
+            synchronized (this) {
+                dataStore = bucketDataMap.get(bucketName);
+                if (dataStore == null) {
+                    dataStore = newStore(bucketName, false);
+                    bucketDataMap.put(bucketName, dataStore);
+                    return;
+                }
+            }
+        }
+    }
+
+    @Override
+    public void dropUnconditionally(String bucketName) throws StorageNodeException {
+        final String TRACE_HEADER = "dropUnconditionally";
+        //We have to handle the following cases:
+        // store is present, connection is open.
+        // store is present, connection is closed.
+        // store is not present.
+
+        DurableLocalStore dataStore = bucketDataMap.get(bucketName);
+        if(dataStore == null) {
+            //open the store if available. 
+            try {
+                load(bucketName);
+            } catch (IllegalStateException e) {
+                //store is deleted already Or is an unknown bucket - nothing to do!
+                logger.info(TRACE_HEADER,"durable store is already deleted on disk / an unknown bucket",
+                            "bucket name", bucketName);
+                return;
+            }
+            dataStore = bucketDataMap.get(bucketName);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `get()` may not be thread-safe at lines: **146, 147, and 157**. In between the check and the `get()` another thread can remove the key and the `get()` will return `null`. The removes that can remove the key are at lines: **166** and **204**.

**Fix**  
 Consider calling `get()`, checking instead of your current check if the returned object is `null`, and then using that object only, without calling `get()` again.

**More info**  
 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3Guard.java#L302-L304) (external link).